packages/themes/src/createTheme.ts

```diff
@@ -1,20 +1,317 @@
+// Support for tuple values in variables like ['#ffffff', '#212126'] (light, dark)
+export type VariableValue = string | readonly [light: string, dark: string];
+
+// Extract all color variable keys from Variables type
+type ColorVariableKeys = {
+  [K in keyof Variables]: K extends `color${string}` ? K : never;
+}[keyof Variables];
+
+// Manually exclude deprecated color variables (marked with @deprecated in Variables type)
+type DeprecatedColorVariables =
+  | 'colorTextOnPrimaryBackground' // @deprecated Use colorPrimaryForeground instead
+  | 'colorText' // @deprecated Use colorForeground instead
+  | 'colorTextSecondary' // @deprecated Use colorMutedForeground instead
+  | 'colorInputText' // @deprecated Use colorInputForeground instead
+  | 'colorInputBackground'; // @deprecated Use colorInput instead
+
+// Get all non-deprecated color variables that should support tuple values
+type NonDeprecatedColorVariables = Exclude<ColorVariableKeys, DeprecatedColorVariables>;
+
+// Extended Variables type that supports tuple values for all non-deprecated color properties
+export type VariablesWithTuples = {
+  [K in keyof Variables]: K extends NonDeprecatedColorVariables ? string | [string, string] : Variables[K];
+};
+
+type ProcessedVariables = {
+  cssString: string | null;
+  processedVariables: Record<string, string | number | undefined>;
+};
+
+type ThemeConfiguration = {
+  variables?: VariablesWithTuples;
+  elements?: Elements | DarkModeElementsFunction;
+  baseTheme?: BaseTheme;
+};
+
+type ThemeOptions = {
+  darkModeSelector?: string | false;
+};
+
+const CAMEL_CASE_REGEX = /[A-Z]/g;
+const LEADING_HYPHEN_REGEX = /^-/;
+const CSS_VAR_PREFIX = '--clerk-';
+const THEME_TYPE = 'prebuilt_appearance';
+const MEDIA_QUERY_PREFIX = '@media';
+const CLASS_SELECTOR_PREFIX = '.';
+const ID_SELECTOR_PREFIX = '#';
+const INTERNAL_GLOBAL_CSS_KEY = '__internal_globalCss';
+
+/**
+ * Converts camelCase to kebab-case for CSS variables
+ * @example toKebabCase('colorBackground') => 'color-background'
+ * @example toKebabCase('XMLParser') => 'x-m-l-parser' (no leading hyphen)
+ */
+export function toKebabCase(str: string): string {
+  return str.replace(CAMEL_CASE_REGEX, letter => `-${letter.toLowerCase()}`).replace(LEADING_HYPHEN_REGEX, '');
+}
+
+/**
+ * Normalizes darkModeSelector with support for opting out
+ * @param selector - The dark mode selector, false to opt out, or undefined for media query fallback
+ * @returns null if opted out, otherwise a valid selector string
+ */
+export function normalizeDarkModeSelector(selector?: string | false): string | null {
+  if (selector === false) {
+    return null; // Opt out of dark mode CSS generation
+  }
+
+  // If no selector provided, opt out by default (light-only theme)
+  // Dark mode should be explicit opt-in behavior
+  if (selector === undefined) {
+    return null;
+  }
+
+  // Use provided selector, trimmed
+  return selector.trim() || null;
+}
```

Comment:
The JSDoc and implementation differ:

> The dark mode selector, false to opt out, or **undefined for media query fallback**

however, we opt-out in the implementation when undefined is provided.